Penalty frictional mortar contact

recuero opened the pull request:
This PR implements the initial approach to penalty frictional mortar contact constraint.

Closes #21329

recuero committed:
Add penalty to frictional mortar contact (#21329)

It includes an initial ironing example with an auxiliary kernel that
interfaces with the mortar user object to output contact quantities, such
as contact pressure, slips, or velocities.

recuero committed:
Add 3D penalty-based pressure

recuero committed:
Add tests with penalty frictional mortar contact

recuero committed:
Add parallel communication for penalty frictional constraint.

We are now moving data from a nodal-based quantity that needs parallel
communication to obtain data from neighboring elements to a distributed
(penalty) distributed constraint that requires having correct parallel
values at the quadrature points for enforcement as generalized forces.

recuero committed:
Add mortar penalty contact to the contact action

recuero committed:
Add documentation for new mortar user objects

recuero committed:
Commit previously missed file

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1464598/) on 33b0447 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24075/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24075/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d9e8e5486ce7388af8fcd32fe0faff8283409afd`

recuero committed:
Clang format

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1465193/) on 5438d8d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24075/site)

This comment will be updated on new commits.

lindsayad commented:
I'm going to unsubscribe for now. Mention me when it's ready for review

recuero committed:
Address parallel issue with penalty mortar in contact action.

The relationship manager didn't get the chance to act if the user chose
mortar_penalty. This is fixed now.

I also removed the tangential velocity output because it gave different
output if distributed meshes were used for a 3D problem. This is strange
since the resulting displacements, tangential pressure, and accumulated
slip all passed. In fact the accumulated slip is computed from the tangential
velocity.

recuero committed:
Address Daniel's comments

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1467696/) on c8d6d5d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24075/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24075/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1542ca05a186584a315c5643fec9329c9d2b5478`

recuero committed:
Add missing gold files for ironing_penalty.

Force another push to kick off testing.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1468753/) on f3520c2 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1542ca</th>
      <th colspan=4>#24075 f3520c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.08%</td>
      <td>85.08%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>86330</td>
      <td>86334</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15144</td>
      <td>15140</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24075/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24075/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>1542ca</th>
      <th colspan=4>#24075 f3520c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.21%</td>
      <td>89.29%</td>
      <td>-0.92%</td>
      <td>79.94%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3271</td>
      <td>3544</td>
      <td>+273</td>
      <td>287</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>355</td>
      <td>425</td>
      <td>+70</td>
      <td>72</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24075/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24075/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24075/coverage/xfem)

</details>

## Warnings

- `contact` new line coverage rate 79.94% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Add real weighted gap output

Increase converage of frictionless mortar penalty with contact action.

recuero committed:
Cleanup

recuero committed:
Attempt to cover velocity parallel communication code

recuero committed:
Remove communicateReals.

I don't believe it's necessary. I initially added it to make sure
that the output was synchronized.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1474257/) on 8190641 : invalidated by @recuero

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1474257/) on 8190641 : invalidated by @recuero

dschwen commented on modules/contact/src/auxkernels/PenaltyMortarUserObjectAux.C:
```diff
@@ -0,0 +1,96 @@
+}
+
+PenaltyMortarUserObjectAux::PenaltyMortarUserObjectAux(const InputParameters & parameters)
+  : AuxKernel(parameters), _contact_quantity(getParam<MooseEnum>("contact_quantity"))
```
```suggestion
  : AuxKernel(parameters), _contact_quantity(getParam<MooseEnum>("contact_quantity").getEnum<ContactQuantity>())
```

dschwen commented on modules/contact/src/auxkernels/PenaltyMortarUserObjectAux.C:
```diff
@@ -0,0 +1,96 @@
+Real
+PenaltyMortarUserObjectAux::computeValue()
+{
+  if (_contact_quantity == ContactQuantity::NORMAL_PRESSURE)
```
looks like a switch statement would be in order here

dschwen commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,353 @@
+PenaltyFrictionUserObject::PenaltyFrictionUserObject(const InputParameters & parameters)
+  : WeightedVelocitiesUserObject(parameters),
+    _penalty(getParam<Real>("penalty")),
+    _penalty_friction(isParamSetByUser("penalty_friction") ? getParam<Real>("penalty_friction")
```
```suggestion
    _penalty_friction(isParamValid("penalty_friction") ? getParam<Real>("penalty_friction")
```

recuero replied:
The parameter is always going to be valid, isn't it? I thought a default value would make it valid. I am interested to check if the user made the effort to set it in the input file.

recuero replied:
I see this comment is linked to the next one.

dschwen commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,353 @@
+  params.addClassDescription(
+      "Computes the mortar frictional contact force via a penalty approach.");
+  params.addRequiredParam<Real>("penalty", "The penalty factor for normal interaction.");
+  params.addParam<Real>("penalty_friction",
+                        0.0,
+                        "The penalty factor for frictional interaction. If not provide, the normal "
```
```suggestion
  params.addParam<Real>("penalty_friction",
                        "The penalty factor for frictional interaction. If not provide, the normal "
```

dschwen commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,353 @@
+
+    if (std::abs(_dof_to_real_tangential_velocity[static_cast<const DofObject *>(node)][0]) >
+        TOLERANCE * TOLERANCE)
+      sign_one =
+          (_dof_to_real_tangential_velocity[static_cast<const DofObject *>(node)][0]) /
+          std::abs(_dof_to_real_tangential_velocity[static_cast<const DofObject *>(node)][0]);
```
```suggestion
      sign_one =
          MathUtils::sign(_dof_to_real_tangential_velocity[static_cast<const DofObject *>(node)][0]);
```

recuero commented on modules/contact/include/auxkernels/PenaltyMortarUserObjectAux.h:
```diff
@@ -0,0 +1,51 @@
+public:
+  static InputParameters validParams();
+
+  /**
+   * Factory constructor, takes parameters so that all derived classes can be built using the same
+   * constructor.
+   */
```
It's the word "same" that shouldn't be around.

recuero committed:
Address Daniel's comments

lindsayad commented on modules/contact/include/userobjects/PenaltyFrictionUserObject.h:
```diff
@@ -0,0 +1,84 @@
+  virtual const VariableTestValue & test() const override;
+  virtual bool constrainedByOwner() const override { return false; }
+
+  /// The normal penalty factor
+  const Real _penalty;
```
It really sucks that we have this pattern where we always have to duplicate code between XXXFrictionUserObject and XXXWeightedGapUserObject. Would it make sense to have `WeightedVelocitiesUserObject` inherit from `T` as opposed to `WeightedGapUserObject`?

lindsayad commented on modules/contact/include/userobjects/WeightedGapUserObject.h:
```diff
@@ -105,6 +102,9 @@ class WeightedGapUserObject : public MortarUserObject
   /// A map from node to weighted gap and normalization (if requested)
   std::unordered_map<const DofObject *, std::pair<ADReal, Real>> _dof_to_weighted_gap;
 
+  /// A map from node to unscaled weighted gap
+  std::unordered_map<const DofObject *, ADReal> _dof_to_real_weighted_gap;
```
So I think that we've set up error checking to ensure that we're always nodal. I think we should replace our `DofObject` map keys by `Node` map keys. This should reduce the number of `static_cast` we do

recuero replied:
Doing this would consequences in other parts of the code. For example, in the parallel communication or other mortar contact objects that have not been migrated to the user object design. I'd suggest we add this as a task to a maintenance issue and try to get to that by the end of the FY.

lindsayad replied:
Ok please add it to a maintenance issue. @dschwen told me this morning you guys would largely like this PR merged as-is

recuero replied:
https://github.com/idaholab/moose/issues/23889#issue-1646302606

I am going to push changes that address most of your comments, but not the couple of them that involve deeper design changes.

lindsayad commented on modules/contact/src/auxkernels/PenaltyMortarUserObjectAux.C:
```diff
@@ -0,0 +1,108 @@
+  switch (_contact_quantity)
+  {
+    case ContactQuantityEnum::NORMAL_PRESSURE:
+      if (dynamic_cast<const PenaltyWeightedGapUserObject *>(_user_object))
+        return static_cast<const PenaltyWeightedGapUserObject *>(_user_object)
+            ->getNormalContactPressure(_current_node);
+      else
+        return static_cast<const PenaltyFrictionUserObject *>(_user_object)
+            ->getNormalContactPressure(_current_node);
```
As mentioned previously, I'd like to just be able to `static_cast` to `WeightedGapUserObject` and have `getNormalContactPressure` available. Consequently I think we should also be able to store `_user_object` as a more specific kind of user object pointer

lindsayad commented on modules/contact/src/userobjects/LMWeightedGapUserObject.C:
```diff
@@ -25,9 +27,12 @@ LMWeightedGapUserObject::validParams()
 LMWeightedGapUserObject::LMWeightedGapUserObject(const InputParameters & parameters)
   : WeightedGapUserObject(parameters), _lm_var(getVar("lm_variable", 0))
 {
-  if (!_lm_var)
+  if (isCoupledConstant("lm_variable"))
```
So `getVar` may return non-null if it is a constant?

lindsayad commented on modules/contact/src/userobjects/LMWeightedVelocitiesUserObject.C:
```diff
@@ -38,8 +40,12 @@ LMWeightedVelocitiesUserObject::LMWeightedVelocitiesUserObject(const InputParame
                                     ? getVar("lm_variable_tangential_two", 0)
                                     : nullptr)
 {
-  if (!_lm_normal_var || !_lm_variable_tangential_one)
-    paramError("lm_variable_normal", "The Lagrange multiplier variables must be actual variables.");
+  if (isCoupledConstant("lm_variable_tangential_one") || isCoupledConstant("lm_variable_normal"))
+    paramError("lm_variable_normal",
+               "The Lagrange multiplier variable must be an actual variable and not a constant.");
+  else if (!_lm_normal_var || !_lm_variable_tangential_one)
+    paramError("lm_variable_normal",
+               "The Lagrange multiplier variables must be provided and be actual variables.");
```
I guess I don't understand the need for this added verbosity. My understanding is that `getVar` should return a `nullptr` if the user has passed a constant in the input file, and so the previous error message should cover that case?

recuero replied:
I intended to address a comment from Daniel to provide more error granularity, see https://github.com/idaholab/moose/pull/23591#discussion_r1167175127. I can revert it.

lindsayad commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,351 @@
+
+  for (auto & map_pr : _dof_to_accumulated_slip)
+  {
+    if (_dof_to_old_normal_pressure[map_pr.first] >
```
Should there always be a key? If so, then I would do `libmesh_map_find` instead of `operator[]`

lindsayad commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,351 @@
+{
+  WeightedVelocitiesUserObject::timestepSetup();
+
+  _dof_to_old_weighted_gap.clear();
```
So if we make `WeightedGapVelocitiesUserObject` inherit from `T`, then we wouldn't need lines like this. It would be really good I think to get to a place where derived classes only have to deal with the data they conceptually introduce. And then if someone catches a bug in penalty weighted gaps, we don't have to remember to fix both `PenaltyWeightedGap` and `PenaltyFrictionUserObject`

lindsayad replied:
In that vein, it would be cool if we could make a lot more data `private`. That would suggest we've made our design even stronger and safer!

recuero committed:
Merge branch 'next' into penalty_frictional_mortar_contact

lindsayad commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,351 @@
+Real
+PenaltyFrictionUserObject::getNormalContactPressure(const Node * const node) const
+{
+  for (auto & map_pr : _dof_to_normal_pressure)
+    if (map_pr.first->id() == node->id())
+      return MetaPhysicL::raw_value(
+          libmesh_map_find(_dof_to_normal_pressure, static_cast<const DofObject *>(map_pr.first)));
```
```suggestion
  if (auto it = _dof_to_normal_pressure.find(_mesh.nodePtr(node->id()));
      it != _dof_to_normal_pressure.end())
    return MetaPhysicL::raw_value(it->second);
```
Let's not do a linear search

recuero replied:
All these comments on not doing linear search should be addressed in the latests commit.

lindsayad commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,351 @@
+PenaltyFrictionUserObject::getFrictionalContactPressure(const Node * const node,
+                                                        const unsigned int component) const
+{
+  for (auto & map_pr : _dof_to_frictional_pressure)
+    if (map_pr.first->id() == node->id())
+      return MetaPhysicL::raw_value(libmesh_map_find(
+          _dof_to_frictional_pressure, static_cast<const DofObject *>(map_pr.first))[component]);
```
As above

lindsayad commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,351 @@
+PenaltyFrictionUserObject::getAccumulatedSlip(const Node * const node,
+                                              const unsigned int component) const
+{
+  for (auto & map_pr : _dof_to_accumulated_slip)
+    if (map_pr.first->id() == node->id())
+    {
+      const auto & slip_x = libmesh_map_find(
+          _dof_to_accumulated_slip, static_cast<const DofObject *>(map_pr.first))[component];
+      return MetaPhysicL::raw_value(slip_x);
+    }
```
As above

lindsayad commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,351 @@
+PenaltyFrictionUserObject::getTangentialVelocity(const Node * const node,
+                                                 const unsigned int component) const
+{
+  for (auto & map_pr : _dof_to_real_tangential_velocity)
+    if (map_pr.first->id() == node->id())
+    {
+      const auto & tan_vel_component =
+          libmesh_map_find(_dof_to_real_tangential_velocity,
+                           static_cast<const DofObject *>(map_pr.first))[component];
+      return MetaPhysicL::raw_value(tan_vel_component);
+    }
```
As above

lindsayad commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -0,0 +1,351 @@
+Real
+PenaltyFrictionUserObject::getNormalWeightedGap(const Node * const node) const
+{
+  for (auto & map_pr : _dof_to_real_weighted_gap)
+    if (map_pr.first->id() == node->id())
+      return MetaPhysicL::raw_value(libmesh_map_find(_dof_to_real_weighted_gap,
+                                                     static_cast<const DofObject *>(map_pr.first)));
```
As above

lindsayad reviewed (changes requested):
16/40 done

lindsayad commented:
Maybe let's investigate having `WeightedGapVelocitiesUserObject` inherit from `T` before I do any more review

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1478087/) on 1e16ccc wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `porous_flow` coverage rate (required: 96.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)

recuero commented on modules/contact/include/userobjects/PenaltyFrictionUserObject.h:
```diff
@@ -0,0 +1,84 @@
+  virtual void reinit() override;
+  virtual void timestepSetup() override;
+
+  Real getNormalContactPressure(const Node * const node) const;
```
I agree that removing those castings is good. But, in order to do so, it seems we'd need to create virtual methods in `WeightedGapUserObject` that do not go with the purpose of the object. For example, we'd need a method for `slip` even though `slip` is unrelated to `WeightedGapUserObject`.

I'll create those getters in the base class.

recuero committed:
Address Alex's comments

recuero commented:
A couple of comments by @lindsayad have not been addressed yet. Just pulling them out of the review thread for visibility: 

- "We have this pattern where we always have to duplicate code between XXXFrictionUserObject and XXXWeightedGapUserObject. Would it make sense to have WeightedVelocitiesUserObject inherit from T as opposed to WeightedGapUserObject?. So if we make WeightedGapVelocitiesUserObject inherit from T, then we wouldn't need lines like this. It would be really good I think to get to a place where derived classes only have to deal with the data they conceptually introduce. And then if someone catches a bug in penalty weighted gaps, we don't have to remember to fix both PenaltyWeightedGap and PenaltyFrictionUserObject. In that vein, it would be cool if we could make a lot more data private. That would suggest we've made our design even stronger and safer!"

- "So I think that we've set up error checking to ensure that we're always nodal. I think we should replace our DofObject map keys by Node map keys. This should reduce the number of static_cast we do" (https://github.com/idaholab/moose/issues/23889#issue-1646302606)

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -14,6 +13,21 @@
+    input = 'cylinder_friction_penalty.i'
+    rel_err = 5.0e-4
+    min_parallel = 4
+    max_parallel = 4
```
Why `max_parallel`?

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -14,6 +13,21 @@
+    min_parallel = 4
+    max_parallel = 4
+    heavy = true
+    min_ad_size = '50'
```
```suggestion
```

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -14,6 +13,21 @@
+    max_parallel = 4
+    heavy = true
+    min_ad_size = '50'
+    requirement = 'Contact module shall approximate frictional contact solution between a cylinder '
```
```suggestion
    requirement = 'The contact module shall approximate frictional contact solution between a cylinder '
```

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -25,6 +39,29 @@
+    abs_zero = 1.0e-4
+    min_ad_size = '200'
+    cli_args = 'Executioner/end_time=0.15'
+    requirement = 'Contact module shall solve frictional contact between a semicircular tool and '
```
```suggestion
    requirement = 'The contact module shall be able to solve frictional contact between a semicircular tool and '
```

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -25,6 +39,29 @@
+    abs_zero = 1.0e-4
+    min_ad_size = '200'
+    cli_args = 'Executioner/end_time=0.15'
+    requirement = 'Contact module shall solve frictional contact between a semicircular tool and '
```
```suggestion
    requirement = 'The contact module shall be able to solve frictional contact between a semicircular tool and '
```

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -49,15 +86,20 @@
+    cli_args = "Constraints/weighted_gap_lm/correct_edge_dropping=true "
+               "UserObjects/weighted_vel_uo/correct_edge_dropping=true Executioner/end_time=5.5 "
+               "Outputs/file_base=frictional_bouncing_block_edge_dropping"
+    requirement = 'Contact module shall solve frictional contact between a bouncing block and a '
```
```suggestion
    requirement = 'The contact module shall be able to solve frictional contact between a bouncing block and a '
```

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -49,15 +86,20 @@
+               "UserObjects/weighted_vel_uo/correct_edge_dropping=true Executioner/end_time=5.5 "
+               "Outputs/file_base=frictional_bouncing_block_edge_dropping"
+    requirement = 'Contact module shall solve frictional contact between a bouncing block and a '
+                  'flexible substrate when correct edge dropping is enabled. Additional requirement '
```
```suggestion
                  'flexible substrate when correct edge dropping is enabled. An additional requirement '
```

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -49,15 +86,20 @@
+               "Outputs/file_base=frictional_bouncing_block_edge_dropping"
+    requirement = 'Contact module shall solve frictional contact between a bouncing block and a '
+                  'flexible substrate when correct edge dropping is enabled. Additional requirement '
+                  'is that the correct edge dropping treatment must yield same results as incorrect '
```
```suggestion
                  'is that the correct edge dropping treatment must yield the same results as an incorrect '
```

lindsayad commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -49,15 +86,20 @@
+    requirement = 'Contact module shall solve frictional contact between a bouncing block and a '
+                  'flexible substrate when correct edge dropping is enabled. Additional requirement '
+                  'is that the correct edge dropping treatment must yield same results as incorrect '
+                  'edge dropping treatment when there is not actual edge dropping, e.g.: All '
```
```suggestion
                  'edge dropping treatment when there is not edge dropping, e.g. when the entire '
```

lindsayad reviewed (changes requested):
Can we delete all the extra code in the constraint classes now?

recuero commented on modules/contact/include/userobjects/WeightedGapUserObject.h:
```diff
@@ -36,6 +36,53 @@ class WeightedGapUserObject : public MortarUserObject
+   * @param component Component of the frictional pressure vector
+   * @return The frictional contact pressure at the node
+   */
+  virtual Real getFrictionalContactPressure(const Node * const /*node*/,
```
> Maybe the aux kernel should be split up? 

@dschwen just mentioned that that's a possibility to tackle this issue. I initially wanted to avoid that, but since more code is being added for augmented Lagrange, I am fine with it.

recuero commented on modules/contact/test/tests/3d-mortar-contact/tests:
```diff
@@ -108,4 +108,26 @@
+                  'constraint using a penalty approach'
+    rel_err = 1.0e-5
+    abs_zero = 1.0e-5
+    max_parallel = 8
```
The reason I initially added that setting is because it has a small number of finite elements, so it may not make sense to run it with a lot of processors. I will remove it.

lindsayad replied:
It may not make sense from a performance standpoint, but it's good for test robustness if it can run with larger processor counts

recuero replied:
Makes sense. 

At this point I am confused as to whether in setting, for this test in particular, has any practical meaning. I know we can set `min_parallel` and force CIVET to employ that number of processors, but without it, wouldn't it just do a two-way run under a parallel module recipe?

lindsayad replied:
You mean `-p2` ? On CIVET this recipe may only be run with 2 processes ... but some day we may add a parallel sweep recipe for modules or contact specifically. Additionally a user may run with something other than `-p2`

recuero replied:
> You mean -p2 ? 

Yes.

recuero commented:
@lindsayad I don't understand this: `Can we delete all the extra code in the constraint classes now?` is it referring to _other_ mortar mechanical objects?

recuero commented on modules/contact/include/userobjects/WeightedGapUserObject.h:
```diff
@@ -36,6 +36,53 @@ class WeightedGapUserObject : public MortarUserObject
+   * @param node Node pointer
+   * @return The normal contact pressure at the node
+   */
+  virtual Real getNormalContactPressure(const Node * const /*node*/) const
```
`LMWeightedGapUserObject` doesn't override this one, so it'd become an abstract class.

lindsayad replied:
But it could, right? It would just return the LM nodal value?

recuero replied:
It could, I think. We have a Lagrange-multiplier formulation when we call `LMWeightedGapUserObject` as opposed to a penalty one. That is, the contact pressure values can be analyzed directly as variable outputs. The auxiliary kernels I am proposing in this PR are meant to leverage the penalty objects. Not that it cannot/should not be changed, but deeper changes would be required. As an example, we don't have accumulated slip in the frictional problem when enforced with Lagrange multipliers.

Let me know if you'd like me to extend that getter to the LM constraint.

lindsayad replied:
Nah it's fine

recuero committed:
Address more of Alex's requests for changes

recuero commented on modules/contact/src/userobjects/WeightedGapUserObject.C:
```diff
@@ -117,15 +116,16 @@ WeightedGapUserObject::computeQpIProperties()
-    _dof_to_weighted_gap[dof].second += (*_test)[_i][_qp] * _qp_factor;
+  _dof_to_weighted_gap[dof].second += (*_test)[_i][_qp] * _qp_factor;
+  _dof_to_real_weighted_gap[dof] =
+      _dof_to_weighted_gap[dof].first / _dof_to_weighted_gap[dof].second;
```
> Do you really want these additional data members as opposed to just doing the operation you're doing here when someone asks for a "real" quantity? 

I removed that data member. However, there are other data members that need to be kept track of. So the same solution doesn't seem to apply. We should profile the code for large problems.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1481136/) on 0e1b6a6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24075/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24075/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 628ea92c69a41195eb6ee8f050c3c696bd48fc6b`

lindsayad commented:
> @lindsayad I don't understand this: Can we delete all the extra code in the constraint classes now? is it referring to other mortar mechanical objects?

Oh it looks like you (or I) already deleted the code in `ComputeWeightedGapLMMechanicalContact` and `ComputeFrictionalForceLMMechanicalContact`. But the Cartesian constraints are still doing their own calculations and using their own data. I suppose that should also be a separate maintenance task as it's distinct from the concepts of this PR

recuero commented:
> But the Cartesian constraints are still doing their own calculations and using their own data. I suppose that should also be a separate maintenance task as it's distinct from the concepts of this PR

I agree. And the dynamic constraints can also leverage the user object design.
